### PR TITLE
More fixes

### PR DESCRIPTION
**Pull request by belldandu**
users was broken and labeled as a dict it is now a list
g_participants now pushes to users and gets its info from _frame or frame.js
participant now pushes to users and gets its info from _frame or frame.js
show_fw now gets its info from _frame or frame.js
tb now gets its info from _frame or frame.js
show_tb now gets its info from _frame or frame.js
premium now gets its info from _frame or frame.js
also fixes have been added and applied to frame.js as well and they have all been tested :D
and are known to work :D 
Bansearchresult completed
blocklist fixed
blocked fixed 
unblocked fixed
premium updated
working on deleted error
fixed problems with recognition of whether the bot was a mod or admin or not 
fixed problems with ban list ban list back to old format and now works 
fixed an issue with spliting arguments that had ; in them in the banlist
fixed delmsg
added delUser
fixed clearUser
fixed flag
added flagUser
fixed ban
fixed unban
fixed problems with message system where everything was undefined >.>
other major fixes which were too numerous to put here 
lib is now at a stable version if you want you can update the version number
:D

### CODE REVIEW

**Comment by belldandu** [comment]
final update tonight i fixed alot of stuff this patch should definetly be merged as both ch_room.js and frame.js need this update

**Comment by MakuraYami** [comment]
Thanks a bunch, great to see someone is working with it :3 i don't have time to read it fully now, took a quick look and as i said before my local version has most of this done and more changes but its not done, and I'm currently too busy with different projects. As soon as I get some time I'll merge my latest code with your fixes

**Comment by belldandu** [comment]
oh
i need one more small fix

**Comment by belldandu** [comment]
there we go i think that should just about do it xD

**Comment by belldandu** [comment]
lol having a few probs with searchban but everything else is working xD

**Comment by belldandu** [comment]
i might make a rooms thing but probably not now since i have no idea how to make the bot leave the room after its joined other than restarting the script

**Comment by belldandu** [comment]
banning still doesn't work but you can check the ban list now and u can get the names of all the users currently in the chat as well as numerous other things so all in all i feel i've done well for a mediocre in javascript although i did learn python first so its not all that different the regular expressions are a pain but meh im getting used to them :P

**Comment by belldandu** [comment]
Last update for tonight I'm tired so yeah XD

File: derplib/frame.js [hunk 1]
```diff
@@ -138,6 +138,44 @@ var frameTypesRoom = {
 			name: name,
 			time: parseFloat(time)};
 	},
+
+	bansearchresult: function(_noidea, name, ip, unid, bansrc, timep1, timep2, timep3){
+		var time = new Date(Date.parse(String(String(timep1.split(' ').join('T'))+':'+String(timep2)+':'+String(timep3)+'.000')));
```
Comment (belldandu) [inline derplib/frame.js]: gg do not remove this

File: derplib/frame.js [hunk 2]
```diff
@@ -138,6 +138,44 @@ var frameTypesRoom = {
 			name: name,
 			time: parseFloat(time)};
 	},
+
+	bansearchresult: function(_noidea, name, ip, unid, bansrc, timep1, timep2, timep3){
```
Comment (belldandu) [inline derplib/frame.js:142]: the time has : in it twice so it tries to split it >.> sadly

File: derplib/frame.js [hunk 3]
```diff
@@ -138,6 +138,44 @@ var frameTypesRoom = {
...
+		var time = new Date(Date.parse(String(String(timep1.split(' ').join('T'))+':'+String(timep2)+':'+String(timep3)+'.000')));
+		return {
+			type: "bansearchresult",
+			name: name,
+			ip: ip,
+			unid: unid,
+			bansrc: bansrc,
+			time: String(time)};
```
Comment (belldandu) [inline derplib/frame.js]: yes it must be a string or its gonna be the date without it being in a string which is gonna cause error's so do not remove

File: derplib/ch_room.js [hunk 4]
```diff
@@ -558,15 +574,15 @@ Room.prototype.clearall = function() {
...
+		if(this.users[user].key && this.users[user].ip && this.users[user].name)
+			this.write(['block', this.users[user].key, this.users[user].ip, this.users[user].name]);
 	}
 }
 
 Room.prototype.unban = function(user) {
 	if(this._isModerator) {
-		if(this._bans[user.name]){
-			this.write(['removeblock', this.bans[user.name].id, this.bans[user.name].ip]); 
+		if(this._bans[user].username){
```
Comment (belldandu) [inline derplib/ch_room.js:602]: username is not a typo do not change

File: derplib/ch_room.js [hunk 5]
```diff
@@ -558,15 +574,15 @@ Room.prototype.clearall = function() {
...
+			this.write(['removeblock', this._bans[user].id, this._bans[user].ip, this._bans[user].username]);
```
Comment (belldandu) [inline derplib/ch_room.js:603]: username is not a typo do not change

**Comment by belldandu** [comment]
lots of stuff fixed
this should definetly be merged at this point xD

**Comment by belldandu** [comment]
im done for a week this should be merged as soon as possible 
every thing is stable for now

**Comment by belldandu** [comment]
i had forgot a ( its fixed now

**Comment by belldandu** [comment]
> .> im tired

**Comment by MakuraYami** [comment]
Ok, ill look at it tomorrow :D

**Comment by belldandu** [comment]
On a side note everything but delete works correctly all I have to do is make a function for it that let's me find the message id using the user name I also need to make msg.deleted the first half of that msg defined cause its not defined the first part ill try to tackle by making messages a object and see what I can do from there but that will most likely be when i have time (either Sunday and Monday or next week) but yeah everything else works :D

**Comment by belldandu** [comment]
like self._messages[user] = [messages] and then var messages;
var message = { user: message.user,
time: message.time,
id: message.id,
?: ?,
?: ?
}
return messages.push(message);
Something along those lines also I did this from phone so it looks crappy but you get the general idea xD

**Comment by belldandu** [comment]
Lol I forgot body: message.body xD

**Comment by belldandu** [comment]
Examples but the concept is easy to understand xD

File: derplib/frame.js [hunk 6]
```diff
@@ -71,7 +71,7 @@ var frameTypesRoom = {
 		return {type: "show_tb", time: seconds};
 	},
 	
-	"delete": function(msgid) {
+	delete: function(msgid) {
```
Comment (MakuraYami) [inline derplib/frame.js]: delete is a js keyword thats why it had quotes. This is bad mkay :(

Comment (belldandu) [inline derplib/frame.js]: mkay ill fix it xD

**Comment by belldandu** [comment]
fixed

**Comment by belldandu** [comment]
there now the lib should be stable :D 
deleted was fixed

**Comment by belldandu** [comment]
that last fix is final fix for this version of derplib no more fixes needed till u add more stuff

**Comment by belldandu** [comment]
also id is set to undefined so that an actual value can be placed there cause false is a boolean and can only be overwritten by true

**Comment by belldandu** [comment]
gg im done for now

**Pull request merged**